### DISCUSSION TITLE
Rect fill not filling completely when scaled

### DISCUSSION THREAD

**Discussion by blakegearin** [fabricjs/fabric.js#8356] [Q&A]
## Version

5.0.0

## Test Case

- [x] [Browser Issue Template](https://jsfiddle.net/Lcp2h3nv/)
- [ ] [Node Issue Template](https://codesandbox.io/s/exciting-browser-ytb701)

https://jsfiddle.net/0bcroum2/1/

## Information about environment

I'm on Firefox `106.0b8`

## Steps To Reproduce

1. Drag the middle right control of the `Rect` to the right edge of the canvas, then continue dragging back to the left

## Expected Behavior

`fill` continuously fills a `Rect` during a scale.

## Actual Behavior

After scaling and making contact with the right edge of the canvas, the `fill` shrinks to the pre-scale `width` with transparent bars on either side. The shrunk `fill` continues to be present until scaling stops, at which point it completely fills the `Rect` again.

## Notes

- I understand this _might_ be expected behavior due to changing `width`, `cacheWidth` and `scaleX` in the middle of a scale. Curious what about this combination makes it behave this way though, and if it can be avoided. Perhaps I'm not considering another relevant attribute

- Goal here was to restrict the `Rect` so it doesn't go off the canvas, so if there's a better pattern for that feel free to share

**Comment by blakegearin** [comment]
@ShaMan123 May I ask why you converted this to a conversation? Are you sure it's not a bug?

**Reply by ShaMan123** [reply]
I labeled as misuse, that is why I made it a discussion
You are misusing the scale control.
Could be worth discussing if the control logic should be more robust to allow such a modification to the object while modifying.

**Reply by ShaMan123** [reply]
https://github.com/fabricjs/fabric.js/blob/master/src/controls/controls.actions.ts#L630

**Reply by ShaMan123** [reply]
If I were you I would consider implementing a control action that keeps the object bound

Though I have to say I agree that the control should be robust enough.
If we change the scaling diff to be calculated from the previous size that should fix this.
OK this might be a bug after all.

**Reply by ShaMan123** [reply]
You can also precalculate the target size so it is kept bound to the viewport and set `lockScalingX/Y` once you reach the size

**Reply by blakegearin** [reply]
I understand. It's not so clear cut.

> If I were you I would consider implementing a control action that keeps the object bound

I'm not familiar with this term, but since posting I have created a working alternative which is to clamp the width after the scaling is finished.

> You can also precalculate the target size

Yes, I have experimented with `lockScalingX` and it can be added to my example with `el.lockScalingX = true;` between L43-46. Though the issue is still present.

> If we change the scaling diff to be calculated from the previous size that should fix this

I was under the impression it was, depending on what you mean by "previous." Based on experimentation I've done it looks like calculations to determine the object's current _actual width during scaling_ is based on `cacheWidth`. 

Here's a detailed outline of what I've observed when changing different element attributes:

| `width` | `cacheWidth` | `scaleX` | Behavior on border | Stops at boundary |
|:-:|:-:|:-:|:--|:-:|
| ✅ |  |  | Transparent bars | ✅ |
|  | ✅ |  | No visual difference | ❌ |
|  |  | ✅ | Snaps element back to the original size | ✅ |
| ✅ | ✅ |  | Transparent bars | ❌ |
| ✅ |  | ✅ | Transparent bars | ✅ |
|  | ✅ | ✅ | Snaps element back to the original size | ✅ |
| ✅ | ✅ | ✅ | Transparent bars | ✅ |

**Note:** I noticed if you change my example to be a `Circle` instead of `Rect` it makes it more obvious that scale is what's responsible for the transparent bars. Since it will go from starting as a circle, then oval, then back to circle again at the border.

**Speculation:** Unfortunately I'm not familiar with this codebase, but if I had to guess, I'm betting the issue is that `cacheWidth` is actually saved in a cache. So manually changing it to a smaller amount causes the next calculation of `scale` (or some behind the scenes property that determines what the _actual_ current width is based on the mouse) to be a negative amount. And since a negative amount isn't expected, the element falls back to the starting values for `width` and `scale`.
